active/InputValidation/InputValidation.md

```diff
@@ -0,0 +1,572 @@
+
+# API Notes
+<!-- Option 1: Give a one or two line description of each API (type
+and member), or at least the ones that aren't obvious
+from their name.  These descriptions are what show up
+in IntelliSense. For properties, specify the default value of the property if it
+isn't the type's default (for example an int-typed property that doesn't default to zero.) -->
+
+<!-- Option 2: Put these descriptions in the below API Details section,
+with a "///" comment above the member or type. -->
+
+# API Details
+<!-- The exact API, in MIDL3 format (https://docs.microsoft.com/en-us/uwp/midl-3/) -->
+
+## InputPropertyAttribute
+The `InputPropertyAttribute` helps us ensure that we only generate validation code for the property specified by this attribute. This will help reduce code bloat by not generating code for properties that will never perform validation (like width). Controls can specify multiple properties as their input property. 
+
+    /// <summary>
+    /// Name of the property that is associated with user input.
+    /// </summary>
+    attribute InputPropertyAttribute
+    {
+         string Name;
+    }
+
+## InputValidationKind  
+Determines how the control displays validation visuals.
+
+    /// <summary>
+    ///    Determines how the control displays it's validation visuals
+    /// </summary>
+    [webhosthidden]
+    enum InputValidationKind
+    {
+        /// <summary>
+        /// Let's the Control dictate how to display visuals. This is the default value.
+        /// </summary>
+        Auto,
+        /// <summary>
+        /// An error icon displays to the right of the control that displays the error messages in a tool tip.
+        /// </summary>
+        Compact,
+        /// <summary>
+        /// Text for the error messages is displayed underneath the control.  
+        /// </summary>
+        Inline,
+        /// <summary>
+        /// Validation visuals are not displayed at all.
+        /// </summary>
+        Hidden,
+    };
+
+## IInputValidationControl
+Any control that wants to participate in input validation should derive from this interface. This, among other things, is what allows the markup compiler to know when to generate the code necessary for listening to the `INotifyDataErrorInfo.ErrorsChanged` event and propagate the results to the `IInputValidationControl.Errors` property.
+
+When the `ValidationCommand` property is non-null, the following will be true:
+1. When the `InputValidationKind` property is set to `Hidden`, the control will *not* invoke the `ValidationCommand.CanValidate` or `ValidationCommand.Validate` methods
+2. When the `InputValidationKind` property is set to `Auto`, the control will use the value of the `ValidationCommand.InputValidationKind` property for the value. If both are `Auto` the default bevahior should be `Inline`.
+3. When focus has been lost, validation should be attempted by calling the `ValidationCommand.CanValidate` method.
```

Comment:
What if `UpdateSourceTrigger` is set to `TextChanged`? - In this case I'd expect the validation to be triggered on TextChanged too. If a develoepr wanted that would they need to invoke validation themselves?
Or will validation always be called on LostFocus, regardless of if it's called elsewhere too?

Reply:
So my goal was to try and decouple the features that come from using `x:Bind` and how validation is designed to work. In my opinion, it is better for us to design input validation independently, so that developers who aren't using the MVVM pattern can use input validation nicely. I still need to validate this API would meet those goals for an app using a framework like ReactiveUI.

As part of the original design, there was an `InputValidationMode` enum that would allow more fine grained control over this. However, my thoughts are that the correct end-user experience for validation is that we should only do initial validation on FocusLost (once user has signaled they are done typing). Then, once the control is invalidated, we try to re-validate on PropertyChanged, so that the user knows as soon as they have fixed the issue. I couldn't really think of a scenario where you wouldn't want this to be the experience for the end-user, so i ommitted the enum from the spec.

The goal here was to describe how validation should work, regardless if `x:Bind` is being used or not.  The only thing that using `x:Bind` really gives you is the boilerplate code to listen to `INotifyDataErrorInfo.ErrorsChanged` and pass those errors to the control. Setting `UpdateSourceTrigger` only controls when the target tries to update the source, and in turn that means your model will raise the `ErrorsChanged` event (assuming invalid input)

Reply:
As it wasn't mentioned explicitly I just wanted to make sure it had been considered and wasn't likely to cause any unforeseen problems if attempted. Thanks for clarifying.